rollup/da_syncer/da_syncer.go

```diff
@@ -14,39 +15,66 @@ var (
-	if block.PartialHeader.Number <= currentBlock.Number().Uint64() {
+	// if override is true, we allow blocks to be lower than the current block number and replace the blocks.
+	if !override && block.PartialHeader.Number <= currentBlock.Number().Uint64() {
 		log.Debug("block number is too low", "block number", block.PartialHeader.Number, "parent block number", currentBlock.Number().Uint64())
 		return ErrBlockTooLow
 	} else if block.PartialHeader.Number > currentBlock.Number().Uint64()+1 {
 		log.Debug("block number is too high", "block number", block.PartialHeader.Number, "parent block number", currentBlock.Number().Uint64())
 		return ErrBlockTooHigh
 	}
 
-	parentBlock := s.blockchain.GetBlockByNumber(currentBlock.Number().Uint64())
+	parentBlockNumber := currentBlock.Number().Uint64()
+	if override {
+		parentBlockNumber = block.PartialHeader.Number - 1
+		// reset the chain head to the parent block so that the new block can be inserted as part of the new canonical chain.
+		err := s.blockchain.SetHead(parentBlockNumber)
+		if err != nil {
+			return fmt.Errorf("failed setting head, number: %d, error: %v", parentBlockNumber, err)
+		}
+	}
+
+	parentBlock := s.blockchain.GetBlockByNumber(parentBlockNumber)
 	if parentBlock == nil {
-		return fmt.Errorf("parent block not found at height %d", currentBlock.Number().Uint64())
+		return fmt.Errorf("failed getting parent block, number: %d", parentBlockNumber)
 	}
 
-	if _, err := s.blockchain.BuildAndWriteBlock(parentBlock, block.PartialHeader.ToHeader(), block.Transactions); err != nil {
+	fullBlock, writeStatus, err := s.blockchain.BuildAndWriteBlock(parentBlock, block.PartialHeader.ToHeader(), block.Transactions, sign)
+	if err != nil {
 		return fmt.Errorf("failed building and writing block, number: %d, error: %v", block.PartialHeader.Number, err)
 	}
+	if writeStatus != core.CanonStatTy {
+		return fmt.Errorf("failed writing block as part of canonical chain, number: %d, status: %d", block.PartialHeader.Number, writeStatus)
+	}
+
+	currentBlock = s.blockchain.CurrentBlock()
+	if currentBlock.Number().Uint64() != fullBlock.NumberU64() || currentBlock.Hash() != fullBlock.Hash() {
+		return fmt.Errorf("failed to insert block: not part of canonical chain, number: %d, hash: %s - canonical: number: %d, hash: %s", fullBlock.NumberU64(), fullBlock.Hash(), currentBlock.Number().Uint64(), currentBlock.Hash())
+	}
+
+	if fullBlock.Number().Uint64()%100 == 0 {
+		log.Info("L1 sync progress", "blockchain height", fullBlock.Number().Uint64(), "block hash", fullBlock.Hash(), "root", fullBlock.Root())
+	}
 
-	if s.blockchain.CurrentBlock().Number().Uint64()%1000 == 0 {
-		log.Info("L1 sync progress", "blockchain height", s.blockchain.CurrentBlock().Number().Uint64(), "block hash", s.blockchain.CurrentBlock().Hash().Hex(), "root", s.blockchain.CurrentBlock().Root().Hex())
+	if s.l2EndBlock > 0 && s.l2EndBlock == block.PartialHeader.Number {
```

Comment:
Do we use `==` here and not `<=` because the whole DA sync process is sequential?

Reply:
Yes the whole DA process is sequential, so we must reach this block number eventually. Specifically for this check we only want it to take effect at exactly this block. this way we allow the user to specify until which L2 block we want to recover